### PR TITLE
DM-18895 proposal: use bigger index types

### PR DESCRIPTION
**Pull request by PaulPrice**
This is a proposal for DM-18895; this fix, combined with a fix to jointcal_cholmod, gets us past the segfault. However, it does introduce extensive changes to the index types (probably more than strictly required) and a dirty hack to get around Eigen's lack of support for these larger index types, so I want to give people a chance to object before I clobber the existing ticket branch.

### CODE REVIEW

File: include/lsst/jointcal/Tripletlist.h [hunk 1]
```diff
@@ -38,16 +38,20 @@ typedef Eigen::Triplet<double> Trip;
 // it would be wise to implement it differently if talking to cholmod
 class TripletList : public std::vector<Trip> {
 public:
+    using Index = Eigen::Index;
```
Comment (parejkoj) [inline include/lsst/jointcal/Tripletlist.h]: Why use this here and not elsewhere (and/or the `vector<Index>` typedef)?

File: include/lsst/jointcal/Tripletlist.h [hunk 2]
```diff
@@ -38,16 +38,20 @@ typedef Eigen::Triplet<double> Trip;
...
+
     TripletList(int count) : _nextFreeIndex(0) { reserve(count); };
 
-    void addTriplet(const unsigned i, const unsigned j, double val) { push_back(Trip(i, j, val)); }
+    void addTriplet(Index const i, Index const j, double val) {
```
Comment (parejkoj) [inline include/lsst/jointcal/Tripletlist.h]: I guess we can get rid of those two primitive `const`s while we're at it.

File: include/lsst/jointcal/Eigenstuff.h [hunk 3]
```diff
@@ -27,12 +27,36 @@
 
 #include "lsst/pex/exceptions.h"
 
+
+#if 1
 #include "Eigen/CholmodSupport"  // to switch to cholmod
+#else
```
[the remark this answers is not in the document]
Comment (PaulPrice) [reply include/lsst/jointcal/Eigenstuff.h]: Oh, I thought I had gotten rid of this block, but I only disabled it. I'll fix that up.

File: include/lsst/jointcal/Eigenstuff.h [hunk 4]
```diff
@@ -72,11 +96,11 @@ class CholmodSimplicialLDLT2
...
         cholmod_sparse *C_cs_perm =
-                cholmod_submatrix(&C_cs, (int *)Base::m_cholmodFactor->Perm, Base::m_cholmodFactor->n,
-                                  nullptr, -1, true, true, &this->cholmod());
+                cholmod_l_submatrix(&C_cs, (Eigen::Index*)Base::m_cholmodFactor->Perm,
+                                  Base::m_cholmodFactor->n, nullptr, -1, true, true, &this->cholmod());
         assert(C_cs_perm);
-        int isOk = cholmod_updown(UpOrDown, C_cs_perm, Base::m_cholmodFactor, &this->cholmod());
-        cholmod_free_sparse(&C_cs_perm, &this->cholmod());
+        int isOk = cholmod_l_updown(UpOrDown, C_cs_perm, Base::m_cholmodFactor, &this->cholmod());
+        cholmod_l_free_sparse(&C_cs_perm, &this->cholmod());
```
[the remark this answers is not in the document]
Comment (PaulPrice) [reply include/lsst/jointcal/Eigenstuff.h]: You get `cholmod_l_*` versions if you build jointcal_cholmod a particular way.
I think the commit message should provide sufficient information about the choice of functions (it specifically mentions `cholmod_l_*`). Oh. I was editing the commit message but never finished that up. Sorry, too many balls in the air. I'll push soon,.

File: include/lsst/jointcal/Associations.h [hunk 5]
```diff
@@ -187,7 +187,7 @@ class Associations {
      *
      * @param[in]  minMeasurements  The minimum number of measuredStars for a FittedStar to be included.
      */
-    void selectFittedStars(int minMeasurements);
+    void selectFittedStars(std::size_t minMeasurements);
```
[the remark this answers is not in the document]
Comment (PaulPrice) [reply include/lsst/jointcal/Associations.h]: As the commit message says, "Perhaps this is more extensive than it strictly needs to be (e.g., polynomial orders don't need the full dynamic range of `std::size_t`) but consistency helps avoid compiler warnings and annoying casts."

I'll try to investigate reverting the few instances you've pointed out and see if they're really necessary, but in general I think it's good to err on the side of `std::size_t` for the sake of consistency.

File: include/lsst/jointcal/Eigenstuff.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (parejkoj) [reply include/lsst/jointcal/Eigenstuff.h]: I still think we want a comment in this file saying why those versions. We don't want to require people to dig up commit messages for such explanations.

File: include/lsst/jointcal/Tripletlist.h (hunk 1 above)
Comment (parejkoj) [reply include/lsst/jointcal/Tripletlist.h]: What about this one? I'm not sure it gains us anything in this file (and I don't think it should be used more broadly than this file), and the `vector<Index>` type takes care of the most complicated use.

**Comment by PaulPrice** [comment]
Replaced by formal submission in #140.

**Pull request closed without merging**